docs/rfcs/0002_templates_mvp.md

```diff
@@ -0,0 +1,359 @@
+| id          | varchar     | A human friendly identifier (Kids First style) for the package |
+| name        | varchar     | Short name for template package                                |
+| description | text        | A short blurb describing the template package                  |
+| created_at  | datetime    | Timestamp of creation                                          |
+| modified_at | datetime    | Timestamp of update                                            |
+| version     | varchar     | Semantic version number of package                             |
+| creator     | foreign key | Foreign key to User that created the template package          |
+
+#### Django Model Schema
+
+| Name                          | Type           | Description                                                                                                      |
+| ----------------------------- | -------------- | ---------------------------------------------------------------------------------------------------------------- |
+| All fields in DB table schema |
+| submission_templates          | RelatedManager | List of templates in this template package                                                                       |
+| package_file                  | FileField      | Reference to the actual package file (archive of the template files in the template package) in a storage system |
+
+## Template Management Rules
+
+For the MVP we can decide on which rules, if any, we want to implement.
+
+Note that if decided not to implement any of the rules, this means it is entirely
+up to the users to ensure templates being used by studies are not
+unintentionally changed or deleted.
+
+-   A SubmissionTemplate may never be updated if it is being used by one or more studies
+
+-   A SubmissionTemplate may never be deleted if it is being used by one or more studies
+
+-   SubmissionTemplates may be added to TemplatePackages at any time
+
+-   SubmissionTemplates may never be deleted from a a TemplatePackage if it
+    is being used by one or more studies.
+
+## ✨ Creating Templates
+
+### CreateSubmissionTemplateMutation
+
+```graphql
+type CreateSubmissionTemplateInput {
+    name: String!
+    dataType: String!
+    description: String!
+    fieldsSchema: JSONString
+    extractConfigFile: Upload
+}
+```
+
+### Permissions
+
+-   Any organization or study administrator may create a template
+
+## Releasing Templates
+
+In order for a template to be used in studies, it must be released/approved
+for use via a `TemplatePackage` and then assigned to a study. The workflow for
+this would be:
+
+1. Create a template package
+    - Add name, description
+    - Add templates
+    - Assign version number
+2. Assign the template package to a study
+
+### CreateTemplatePackageMutation
+
+Creates a TemplatePackage and optionally assigns it to a study
+
+```graphql
+type CreateTemplatePackageInput {
```

Comment:
Hmm I wasn't sure if we should have an organization tied to the template package as well. My idea was that an organization owns and authors a template. And then someone can create a template package with templates that could come from more than one org. 

I wonder what would the purpose of a foreign key to org be on the package? Would that imply that there are additional restrictions to updating the package?

Reply:
So then there would be a super directory for any user/admin in any org to browse through and try to incorporate into one of their studies? There's a lot of use cases to unwrap with that.

Reply:
Yea probably right. Ok so a template package also has an owning org like templates do. And then the update/delete ops for packages will be allowed for org admins that are from the owning org only. No one from outside the org can update/delete packages.

Reply:
Do you think it may be similar to how Cavatica apps work? Maybe easier to 'clone' them into an org?
This sort of thing is easy to implement but I'm afraid there's so many edge cases we are gonna hit.

- Content moderation (make sure things have good names/descriptions)
- Allow things to be sharable/private
- Build out directory/template browser
- Deal with things like the owning org removing a template other's use
- People wanting to take a template and modify it for their needs

Those are just a few I can think of right now.

Reply:
Ok, I updated the MVP. I think this should work decently for an MVP. It alleviates some of those ^ issues via template versions. The longer term solution (if we choose to implement) incorporates package versioning which just makes it easier to manage templates in a logically related group and provides an avenue for a release process.

Reply:
I also wrote up a use case and solutions doc that outlines some use cases and walks through potential solutions for the most common one (modify a template that's being used by others)